### PR TITLE
Implemented Primitive Domain Socket Support described in https://github.com/v2ray/Planning/issues/25

### PR DESCRIPTION
**Pull request by xiaokangwang**
Currently, only basic support for stream unix domain socket is included.

Works:

Dial & Listen for stream unix domain socket.

Won't work:

- Dial/Listen non-stream unix domain socket
- Security for domain socket

### CODE REVIEW

**Comment by xiaokangwang** [comment]
I might have pulled from 2b675335bd0e3845df43d0037d90bddafd588d60 which have test failures. If this is the reason for test fail in this pr, I can add anther merge command to resolve this issue given the issue on master have resolved.

File: transport/internet/domainsocket/listener.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+			close(ls.listenerChan)
+			return
+		}
+		conn, err := ls.ln.Accept()
+
+		if err != nil {
+			newError("Cannot Accept socket from listener").Base(err).AtDebug().WriteToLog()
+			if err == lasterror {
+				errortolerance--
+				if errortolerance == 0 {
```
Comment (xiaokangwang) [inline transport/internet/domainsocket/listener.go]: Guard against too many open file error

File: transport/internet/domainsocket/listener.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+	if !ls.state.Has(STATE_INITIALIZED) || ls.state.Has(STATE_LOWERUP) {
+		err = newError(ls.state).Base(newError("Invalid State:LowerUP"))
+		if ___DEBUG_PANIC_WHEN_ENCOUNTED_IMPOSSIBLE_ERROR {
+			panic(err)
+		}
+		return err
+	}
+
+	if isUnixDomainSocketFileSystemBased(ls.path) && !___DEBUG_IGNORE_FLOCK {
```
Comment (xiaokangwang) [inline transport/internet/domainsocket/listener.go]: If the unix domain socket is filesystem based, an file lock must be used to claim the right for listening on respective file.

Comment (xiaokangwang) [reply transport/internet/domainsocket/listener.go]: https://gavv.github.io/blog/unix-socket-reuse/

File: transport/internet/domainsocket/listener.go (hunk 1 above)
Comment (DarienRaymond) [reply transport/internet/domainsocket/listener.go]: Please put this comment into source code.

File: transport/internet/domainsocket/listener.go (hunk 2 above)
Comment (DarienRaymond) [reply transport/internet/domainsocket/listener.go]: Please put this comment into source code.

**Pull request merged**